z/z.go

```diff
@@ -18,22 +18,47 @@ package z
 
 import (
 	"context"
+	"errors"
+	"log"
 	"sync"
 
 	"github.com/cespare/xxhash/v2"
 )
 
-// TODO: Figure out a way to re-use memhash for the second uint64 hash,
-// we already know that appending bytes isn't reliable for generating a
-// second hash (see Ristretto PR #88).
-// We also know that while the Go runtime has a runtime memhash128
-// function, it's not possible to use it to generate [2]uint64 or
-// anything resembling a 128bit hash, even though that's exactly what
-// we need in this situation.
-func KeyToHash(key interface{}) (uint64, uint64) {
-	if key == nil {
-		return 0, 0
+// GetKeyToHash will get the default KeyToHash function for the primitive types.
+func GetKeyToHash[K any](emptyKey K) (func(K) (uint64, uint64), error) {
+	keyAsAny := any(emptyKey)
+
+	if keyAsAny == nil {
+		log.Fatal("must provide custom KeyToHash function for pointer type")
+	}
+	switch keyAsAny.(type) {
+	case uint64:
+	case string:
+	case []byte:
+	case byte:
+	case int:
+	case int32:
+	case uint32:
+	case int64:
+	default:
+		return nil, errors.New("must provide custom KeyToHash function for given type")
 	}
+
+	return KeyToHash[K], nil
+}
+
+// TODO: Figure out a way to re-use memhash for the second uint64 hash, we
+//
+//	already know that appending bytes isn't reliable for generating a
```

Comment:
you have to fix the double space after the // so that the whole TODO is together.
Additionally, I am thinking we don't need to change anything in the file. The original function works as is?

Reply:
so two things about the original function.
1: we must wrap it with this funciton. 
`func KeyToHash[K any](key K) (uint64, uint64) {
	return defaultKeyToHash(key)
}`
2: i want to catch also on initialization if there is no function given for other types and thats why i want this function added.
`func GetKeyToHash[K any](emptyKey K) (func(K) (uint64, uint64), error)`
but i can theoretically remove this.